Adding decision record for additional Metrics fields (#3313)
spiceai/spiceai: docs/decisions/003-additional-metrics-fields.md

# DR: Adding additional fields to the Anonymous Metrics

## Context

> Spice reports certain metrics anonymously, such as query count, bytes processed, bytes returned, query duration,

Reviewer: The context doesn't address why this DR was created. Why are we pondering this decision?

and query execution duration. 

## Assumptions

1. Adding the new fields does not impose a significant performance impact on Spice

## Options

## First-Principles

> - **Bring data and ML to your application**: Better data means better decisions and a better user experience

Reviewer: It's not clear to me how this is related to data and ML in user applications?

## Decision

Spice will add the following new fields to the metrics:

* `user-agent`: The originating User agent of the request. For example, SDKs report values such as `gospice` for the Go SDK or `spicepy` for the Python SDK
* `user-agent-version`: The version of the originating library
* `user-agent-os`: The operating system and architecture of the originating request
* `spice-internal`: An optional field denoting whether this metric is being generated as part of any Spice internal benchmarks or tests

> * `build`: `dev` or `release` denoting if the current runtime is built in Release mode or not. 

Reviewer: ```suggestion
* `build`: `debug` or `release` denoting if the current runtime is built in Release mode or not. 
```

Maybe this should be debug instead of dev?

**Why**:

> - As Spice grows and adds new features, the need to have more visibility into usage becomes more beneficial.

Reviewer: ```suggestion
- As Spice expands, greater visibility into usage becomes increasingly valuable.
```

> - Being able to separate internal Spice benchmarks versus actual user consumption helps to understand growth

Reviewer: ```suggestion
- Distinguishing internal benchmark/testing from actual user activity clarifies growth.
```

> - Understanding which SDKs are used most will allow us to focus our efforts.

Reviewer: ```suggestion
- Identifying popular SDKs enables focused investment.
```

> - Any performance regressions specific to certain SDKs will be easier to identify.

Reviewer: ```suggestion
- Performance regressions specific to SDKs will be easier to spot.
```

**Why not**:

> There may be performance overhead when adding these new metrics, however small. There is also the concern about malformed or incorrect user agent strings, which should be accounted for

Reviewer: Can we quantify the performance overhead somehow? Also what is the concern about malformed user agent strings?
